### PR TITLE
Add analytics for Xsearch requests

### PR DESCRIPTION
**Pull request by kderusso**
Initial analytics for the REX demo. This requires manually creating and setting up the data stream. You can use the following queries to see analytics: 
```
# Create a component template for mappings
PUT _component_template/acme-analytic-mappings
{
  "template": {
    "mappings": {
      "properties": {
        "@timestamp": {
          "type": "date",
          "format": "date_optional_time||epoch_millis"
        },
        "query": {
          "type": "keyword"
        }
      }
    }
  },
  "_meta": {
    "description": "Mappings for @timestamp and query fields"
  }
}

# Create a template for the data stream
PUT _index_template/acme-analytic-template
{
  "index_patterns": ["acme-analytics*"],
  "data_stream": { },
  "composed_of": [ "acme-analytic-mappings" ],
  "priority": 500,
  "_meta": {
    "description": "Template for ACME analytics time series data"
  }
}

# Create the analytics data stream 
PUT /_data_stream/acme-analytics

# Updated engine creation call, that takes an analytics collection as an argument 
PUT _search_engine/acme-search
{
  "indices": ["search-parks", "search-more-parks"],
  "relevance_settings": "parks-search-configuration",
  "analytics_collection": "acme-analytics"
}

#XSearch is essentially just a passthru to _search here
GET acme-search/_xsearch?explain=true
{
  "query": "forested park"
}

# View the analytics events 
GET /acme-analytics/_search
```

### CODE REVIEW

**Comment by dnhatn** [comment]
@kderusso Should the enterprise search team use a forked repository instead? /cc @qhoxie

**Comment by kderusso** [comment]
@dnhatn We made the decision to use a branch for the purposes of our demo, with the agreement that it won't be merged into `main`. With that said, going forward we'll try to remove the `needs-triage` tag.

File: x-pack/plugin/relevance-search/src/main/java/org/elasticsearch/xpack/relevancesearch/xsearch/analytics/XSearchAnalyticsService.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    private static final Logger logger = LogManager.getLogger(XSearchAnalyticsService.class);
+
+    private final ClusterService clusterService;
+
+    public XSearchAnalyticsService(ClusterService clusterService) {
+        this.clusterService = clusterService;
+    }
+
+    public void recordEvent(XSearchAction.Request request, NodeClient client) {
+        Collection<SearchEngine> engines = getEnginesForRequest(request);
```
Comment (afoucret) [inline x-pack/plugin/relevance-search/src/main/java/org/elasticsearch/xpack/relevancesearch/xsearch/analytics/XSearchAnalyticsService.java:46]: Not something for now, but I do not think it is a good idea to allow to query several engines in a single _xsearch request.

**Review by afoucret — CHANGES_REQUESTED** [review]
Overall great.
Few changes requested, but nothing important.

File: x-pack/plugin/relevance-search/src/main/java/org/elasticsearch/xpack/relevancesearch/xsearch/action/rest/RestXSearchAction.java [hunk 2]
```diff
@@ -68,7 +72,19 @@ public RestChannelConsumer prepareRequest(final RestRequest request, final NodeC
...
+
+    private void performRequest(
+        String index,
+        XSearchAction.Request xsearchRequest,
+        RelevanceMatchQueryBuilder queryBuilder,
+        NodeClient client,
+        RestChannel channel
+    ) {
+        doXSearch(index, xsearchRequest.explain(), queryBuilder, client, channel);
+        xSearchAnalyticsService.recordEvent(xsearchRequest, client);
```
[the remark this answers is not in the document]
Comment (kderusso) [reply x-pack/plugin/relevance-search/src/main/java/org/elasticsearch/xpack/relevancesearch/xsearch/action/rest/RestXSearchAction.java:87]: @afoucret If I remove this, calls to the REST handler don't record analytics.

File: x-pack/plugin/search-engines/src/main/java/org/elasticsearch/searchengines/action/PutSearchEngineAction.java [hunk 3]
```diff
@@ -85,6 +90,18 @@ public void setRelevanceSettingsId(String relevanceSettingsId) {
             this.relevanceSettingsId = relevanceSettingsId;
         }
 
+        public boolean shouldRecordAnalytics() {
```
[the remark this answers is not in the document]
Comment (kderusso) [reply x-pack/plugin/search-engines/src/main/java/org/elasticsearch/searchengines/action/PutSearchEngineAction.java:93]: Good catch, thank you!

File: x-pack/plugin/relevance-search/src/main/java/org/elasticsearch/xpack/relevancesearch/xsearch/action/XSearchTransportAction.java [hunk 4]
```diff
@@ -59,6 +61,8 @@ protected void doExecute(Task task, XSearchAction.Request request, ActionListene
                 l.onResponse(searchResponse);
             } catch (Exception t) {
                 l.onFailure(t);
+            } finally {
+                xSearchAnalyticsService.recordEvent(request, client);
```
Comment (carlosdelest) [inline x-pack/plugin/relevance-search/src/main/java/org/elasticsearch/xpack/relevancesearch/xsearch/action/XSearchTransportAction.java:65]: Nit: Maybe we could add the response as well to allow for storing response related analytics (like number of hits, time spent, etc).

File: x-pack/plugin/relevance-search/src/main/java/org/elasticsearch/xpack/relevancesearch/xsearch/analytics/XSearchAnalyticsService.java [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+        Collection<String> dataStreams = getDataStreamsForEngines(engines);
+        dataStreams.forEach(ds -> recordEvent(ds, request, client));
+    }
+
+    private void recordEvent(String dataStream, XSearchAction.Request request, NodeClient client) {
+        try (XContentBuilder builder = buildEvent(request)) {
+            if (builder != null) {
```
Comment (carlosdelest) [inline x-pack/plugin/relevance-search/src/main/java/org/elasticsearch/xpack/relevancesearch/xsearch/analytics/XSearchAnalyticsService.java:53]: Question for the future: Would there be a way for this to be asynchronous, so we don't block the query response while analytics is being stored? 🤔

**Review by carlosdelest — COMMENTED** [review]
Agree with @afoucret comments about cleanup. The rest LGTM 👍

File: x-pack/plugin/relevance-search/src/main/java/org/elasticsearch/xpack/relevancesearch/xsearch/analytics/XSearchAnalyticsService.java (hunk 5 above)
Comment (afoucret) [reply x-pack/plugin/relevance-search/src/main/java/org/elasticsearch/xpack/relevancesearch/xsearch/analytics/XSearchAnalyticsService.java:53]: @carlosdelest Actually it is asynchronous !!!
A nit would be to add a listener to the method signature and to wrap error logging.

**Review by afoucret — APPROVED** [review]

File: x-pack/plugin/relevance-search/src/main/java/org/elasticsearch/xpack/relevancesearch/xsearch/action/XSearchTransportAction.java (hunk 4 above)
Comment (kderusso) [reply x-pack/plugin/relevance-search/src/main/java/org/elasticsearch/xpack/relevancesearch/xsearch/action/XSearchTransportAction.java:65]: I like that suggestion, but I'd like to do it in a subsequent PR if you're good with that!

File: x-pack/plugin/relevance-search/src/main/java/org/elasticsearch/xpack/relevancesearch/xsearch/action/rest/RestXSearchAction.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (kderusso) [reply x-pack/plugin/relevance-search/src/main/java/org/elasticsearch/xpack/relevancesearch/xsearch/action/rest/RestXSearchAction.java]: I found the issue and fixed! 👍

File: x-pack/plugin/relevance-search/src/main/java/org/elasticsearch/xpack/relevancesearch/xsearch/action/rest/RestXSearchAction.java [hunk 6]
```diff
@@ -36,15 +37,18 @@ public class RestXSearchAction extends BaseRestHandler {
...
     private final RelevanceMatchQueryRewriter relevanceMatchQueryRewriter;
     private final XSearchRequestValidationService xSearchRequestValidationService;
+    private final XSearchAnalyticsService xSearchAnalyticsService;
 
     @Inject
     public RestXSearchAction(
         RelevanceMatchQueryRewriter relevanceMatchQueryRewriter,
-        XSearchRequestValidationService xSearchRequestValidationService
+        XSearchRequestValidationService xSearchRequestValidationService,
+        XSearchAnalyticsService xSearchAnalyticsService
```
[the remark this answers is not in the document]
Comment (kderusso) [reply x-pack/plugin/relevance-search/src/main/java/org/elasticsearch/xpack/relevancesearch/xsearch/action/rest/RestXSearchAction.java]: ✅

**Pull request merged**